src/lib/UniversalDiscovery/CotfPermissionChecker.php

```diff
@@ -0,0 +1,305 @@
+        ContentService $contentService,
+        ContentTypeService $contentTypeService,
+        PermissionResolver $permissionResolver,
+        LimitationService $limitationService
+    ) {
+        $this->locationService = $locationService;
+        $this->contentService = $contentService;
+        $this->contentTypeService = $contentTypeService;
+        $this->permissionResolver = $permissionResolver;
+        $this->limitationService = $limitationService;
+    }
+
+    /**
+     * @param Location $location
+     *
+     * @return \EzSystems\EzPlatformAdminUi\UniversalDiscovery\CotfCreateRestrictions
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\BadStateException
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     */
+    public function getCreateRestrictions(Location $location): CotfCreateRestrictions
+    {
+        $restrictedContentTypesIds = [];
+        $restrictedLanguagesCodes = [];
+
+        $contentCreateRestrictions = $this->resolvePermissionSets('content', 'create', $location);
+        $contentPublishRestrictions = $this->resolvePermissionSets('content', 'publish', $location);
+
+        $hasAccess = $contentCreateRestrictions->hasAccess && $contentPublishRestrictions->hasAccess;
+
+        if ($hasAccess) {
+            $restrictedContentTypesIds = $this->resolveRestrictions(
+                $contentCreateRestrictions->restrictedContentTypesIds,
+                $contentPublishRestrictions->restrictedContentTypesIds
+            );
+
+            $restrictedLanguagesCodes = $this->resolveRestrictions(
+                $contentCreateRestrictions->restrictedLanguagesCodes,
+                $contentPublishRestrictions->restrictedLanguagesCodes
+            );
+        }
+
+        return new CotfCreateRestrictions($hasAccess, $restrictedContentTypesIds, $restrictedLanguagesCodes);
+    }
+
+    /**
+     * @param string $module
+     * @param string $function
+     * @param \eZ\Publish\API\Repository\Values\Content\Location $location
+     *
+     * @return \EzSystems\EzPlatformAdminUi\UniversalDiscovery\CotfCreateRestrictions
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\BadStateException
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     */
+    public function resolvePermissionSets(
+        string $module,
+        string $function,
+        Location $location
+    ): CotfCreateRestrictions {
+        $permissionSets = $this->permissionResolver->hasAccess($module, $function);
+
+        if (\is_bool($permissionSets)) {
+            return new CotfCreateRestrictions($permissionSets);
+        }
+
+        $restrictedContentTypesIds = [];
+        $restrictedLanguagesCodes = [];
+        $locationCreateStruct = $this->locationService->newLocationCreateStruct($location->id);
+        $currentUserRef = $this->permissionResolver->getCurrentUserReference();
+
+        foreach ($permissionSets as $permissionSet) {
```

Comment:
Exposing internal business logic of permission system is at least risky. If the internal logic changes, there won't be a way to track it here. It also might not fully cover extensibility of permission system.